libbpf-tools/syscount.bpf.c

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Anton Protopopov
+//
+// Based on syscount(8) from BCC by Sasha Goldshtein
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "syscount.h"
+
+const volatile bool count_by_process = false;
+const volatile bool measure_latency = false;
+const volatile bool filter_failed = false;
+const volatile int filter_errno = false;
+const volatile pid_t filter_pid = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u64);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, struct data_t);
+} data SEC(".maps");
+
+static __always_inline
+void *bpf_map_lookup_or_try_init(void *map, void *key, void *init)
+{
+	void *val;
+	int err;
+
+	val = bpf_map_lookup_elem(map, key);
+	if (val)
+		return val;
+
+	err = bpf_map_update_elem(map, key, init, 0);
+	if (err)
+		return (void *) 0;
+
+	return bpf_map_lookup_elem(map, key);
+}
+
+SEC("tracepoint/raw_syscalls/sys_enter")
+int tracepoint__raw_syscalls__sys_enter(struct trace_event_raw_sys_enter *args)
+{
+	u64 id;
+	u64 ts;
+
+	/* Currently there is no "skel" way to skip load/attach
+	 * if !measure_latency, so check it here */
+	if (!measure_latency)
+		return 0;
+
+	id = bpf_get_current_pid_tgid();
+	if (filter_pid && id >> 32 != filter_pid)
+		return 0;
+
+	ts = bpf_ktime_get_ns();
+	bpf_map_update_elem(&start, &id, &ts, 0);
+	return 0;
+}
+
+SEC("tracepoint/raw_syscalls/sys_exit")
+int tracepoint__raw_syscalls__sys_exit(struct trace_event_raw_sys_exit *args)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	struct data_t zero = {};
```

Comment:
this is used for initialization only, right? So it's better to make it a static variable and skip zero-initializing it on every call